solr/modules/monitor/src/java/org/apache/solr/monitor/update/MonitorUpdateRequestProcessor.java

```diff
@@ -0,0 +1,257 @@
+  }
+
+  @Override
+  public void processAdd(AddUpdateCommand cmd) throws IOException {
+    var solrInputDocument = cmd.getSolrInputDocument();
+    var queryId =
+        (String) solrInputDocument.getFieldValue(indexSchema.getUniqueKeyField().getName());
+    var queryFieldValue = solrInputDocument.getFieldValue(MonitorFields.MONITOR_QUERY);
+    if (queryFieldValue != null) {
+      var payload =
+          Optional.ofNullable(solrInputDocument.getFieldValue(MonitorFields.PAYLOAD))
+              .map(Object::toString)
+              .orElse(null);
+      List<SolrInputDocument> children =
+          Optional.of(queryFieldValue)
+              .filter(String.class::isInstance)
+              .map(String.class::cast)
+              .map(
+                  queryStr ->
+                      new MonitorQuery(
+                          queryId, SimpleQueryParser.parse(queryStr, core), queryStr, Map.of()))
+              .stream()
+              .flatMap(monitorQuery -> decompose(monitorQuery, payload))
+              .map(this::toSolrInputDoc)
+              .collect(Collectors.toList());
+      if (children.isEmpty()) {
+        throw new SolrException(
+            SolrException.ErrorCode.INVALID_STATE, "Query could not be decomposed");
+      }
+      SolrInputDocument firstChild = children.get(0);
+      if (solrInputDocument.hasChildDocuments()) {
+        solrInputDocument.getChildDocuments().clear();
+      }
+      solrInputDocument.addChildDocuments(children.stream().skip(1).collect(Collectors.toList()));
+      if (solrInputDocument.hasChildDocuments()) {
+        solrInputDocument
+            .getChildDocuments()
+            .forEach(
+                child ->
+                    solrInputDocument.forEach(
+                        field -> {
+                          if (!MonitorFields.RESERVED_MONITOR_FIELDS.contains(field.getName())) {
+                            child.addField(field.getName(), field.getValue());
+                          }
+                        }));
+        solrInputDocument
+            .getChildDocuments()
+            .forEach(
+                child ->
+                    child.setField(
+                        indexSchema.getUniqueKeyField().getName(),
+                        child.getFieldValue(MonitorFields.CACHE_ID)));
+      }
+      copyFirstChildToParent(solrInputDocument, firstChild);
+    }
+    super.processAdd(cmd);
+  }
+
+  private void copyFirstChildToParent(SolrInputDocument parent, SolrInputDocument firstChild) {
+    parent.setField(
+        indexSchema.getUniqueKeyField().getName(),
+        firstChild.getFieldValue(MonitorFields.CACHE_ID));
+    for (var firstBornInputField : firstChild) {
+      parent.setField(firstBornInputField.getName(), firstBornInputField.getValue());
+    }
+  }
+
+  private Stream<Document> decompose(MonitorQuery monitorQuery, String payload) {
+    return QCEVisitor.decompose(monitorQuery, MonitorConstants.QUERY_DECOMPOSER).stream()
```

Comment:
`MonitorConstants.QUERY_DECOMPOSER` being a non-string constant jumps out a little and so I was wondering if a decomposer might live alongside the presearcher perhaps. https://github.com/cpoerschke/solr/commit/6d4986cd20fd91729e146d1f1bd409269c64bb10 commit on https://github.com/cpoerschke/solr/commits/solr-monitor-cpoerschke-2 branch explores that -- tests pass but it's a bit awkward w.r.t. cache use and the query-decoder/query-decomposer classes (and I haven't yet looked into what the query decoder does) but sharing as-is anyhow.

Reply:
I like that you make `SolrMonitorQueryDecoder` a class. The fact that it is an interface is a bit of lazy cargo-cult programming by me copying a luwak pattern. As you said, the fact that your query decomposer lives with the decoder without the decoder using it is awkward but if you can refactor your branch a bit to remove the need for `SolrMonitorQueryCollector` to separately reference the decomposer when it calls `QueryVisitor::getComponents), that might be nice. That's probably where you were going with that? You might need to have `getComponents` lazily accept the decoder and the do the actual decoding within `QueryVisitor` to gain visibility.

Reply:
Tangentially, a drawback to this approach that I haven't called out is that you deserialize the same query many times for each component that gets extracted. I initially thought about structuring the cache in a different way to store a whole set of sibling queries as one entry but I couldn't easily come up with how to make this compatible with the incremental, version-reliant cache update logic currently in place. Basically, I didn't want to assume that sibling+parent doc versions would all be the same and the logic currently relies on version to decide whether or not to update an entry.

Reply:
> I like that you make `SolrMonitorQueryDecoder` a class. ... That's probably where you were going with that? You might ... want to just move `getComponent` from `QCEVisitor` to `SolrMonitorQueryDecoder`? ...

Thanks for taking a look! No, I wasn't actually going anywhere specific with that, the branch was more an exploratory meandering to better understand how the pieces all fit together and how they might fit together differently too.

Thanks for the move suggestion: https://github.com/cpoerschke/solr/commits/solr-monitor-cpoerschke-2 now also contains https://github.com/cpoerschke/solr/commit/0915ce18a4b43810465d9eb5008faf979a53cbf7 with the suggested move and tentatively https://github.com/cpoerschke/solr/commit/e6e25a54abdfdbf754802d80e9235b390a070d63 to avoid the `queryDecoder.getComponent(queryDecoder.decode(...), ...)` usage pattern.

Reply:
Yea I like `https://github.com/cpoerschke/solr/commit/0915ce18a4b43810465d9eb5008faf979a53cbf7` .. one thing I'd add is that it might be even nicer to add a `getComponent(MonitorDataValues values, String cacheId)` that decodes and returns the desired component in one call instead of feeding one output of the decoder back to itself.

Reply:
> ... one thing I'd add is that it might be even nicer to add a `getComponent(MonitorDataValues values, String cacheId)` that decodes and returns the desired component in one call instead of feeding one output of the decoder back to itself.

Yep, that's what the https://github.com/cpoerschke/solr/commit/e6e25a54abdfdbf754802d80e9235b390a070d63 commit after that already does.

Reply:
Both commits from 'there' now merged back into the branch 'here' ...

... leaving the possible `public class SolrMonitorQueryDecoder extends LuceneMonitorQueryDecoder {` exploration for another time, feel free to take that up if you wish and have bandwidth.

Reply:
Where would you envision `LuceneMonitorQueryDecoder ` would live, in lucene itself? My hope was to upstream the small changes I have in that "lucene-monitor package visitor". The changes are really about exposing some of those bits in lucene-monitor. I was planning to write the dev-list there with my argument about why exposing certain methods/classes as `public` is probably for the best. Lucene-monitor really isn't your typical lucene package .. it does a lot of inflexible/unconfigurable state management and I've talked to a few heavy users that complain about it (mainly the all-or-nothing `ConcurrentHashMap` query cache and coarse locking). It is basically a self-contained database with a CRUD API that doesn't mesh well with other databases that might want to use its otherwise really great algorithms. I think this is one of the reasons elastic's percolator never adopted lucene-monitor and has decided to manage their own [lucene-monitor-like functionality](https://github.com/elastic/elasticsearch/blob/main/modules/percolator/src/main/java/org/elasticsearch/percolator/QueryAnalyzer.java).

Reply:
> Where would you envision `LuceneMonitorQueryDecoder ` would live, in lucene itself? ...

Hmm, I don't feel that I've developed enough insights here as yet to have a meaningful opinion (w.r.t. what could live in Lucene vs. living elsewhere vs. changes to Lucene) - sorry.

You mentioned about the decoder calling `QueryCacheEntry` and that made sense to me since the method is currently static in `QCEVisitor` but only called from the decoder (and allocating `QCEVisitor` objects).